File: src/engine/editCommands.ts
import type { LayoutVariant, PlanObject, Plot, Transform } from '../domain/types';
import { OBJECT_LIBRARY } from '../domain/objectLibrary';
import { polygonBounds, rectFullyInsidePolygon, transformAabb, aabbOverlap, distance } from './geometry';

// Bounded-vocabulary editing commands, in the same spirit as textParser.ts's
// brief extraction: free text is matched against a fixed set of verbs and the
// object type labels actually present in the current plan, rather than sent
// to an open-ended model, so results stay deterministic and explainable.

export type EditVerb = 'move-near' | 'move-away' | 'enlarge' | 'shrink' | 'delete' | 'duplicate' | 'rotate' | 'lock' | 'unlock';

export interface ParsedEditCommand {
  verb: EditVerb;
  subjectTypeId: string;
  referenceTypeId?: string;
}

// Order matters: more specific phrases (e.g. "away from") must be tested
// before their broader counterparts wouldn't otherwise conflict, but this
// also fixes priority when a command could ambiguously match more than one
// verb family (delete/duplicate/rotate/lock take priority over resize or
// move phrasing that might incidentally share a word).
const VERB_PATTERNS: { verb: EditVerb; re: RegExp }[] = [
  { verb: 'delete', re: /\b(delete|remove)\b/ },
  { verb: 'duplicate', re: /\b(duplicate|copy|clone)\b/ },
  { verb: 'rotate', re: /\b(rotate|turn)\b/ },
  { verb: 'unlock', re: /\bunlock\b/ },
  { verb: 'lock', re: /\block\b/ },
  { verb: 'enlarge', re: /\b(bigger|larger|enlarge|increase|grow|expand)\b/ },
  { verb: 'shrink', re: /\b(smaller|shrink|decrease|reduce|shrink)\b/ },
  { verb: 'move-away', re: /\b(away from|farther from|further from|far from)\b/ },
  { verb: 'move-near', re: /\b(near|closer to|next to|beside|by the)\b/ },
];

function findVerb(text: string): EditVerb | null {
  for (const { verb, re } of VERB_PATTERNS) {
    if (re.test(text)) return verb;
  }
  return null;
}

// Longest-label-first so a multi-word label (e.g. "Solar Panels") wins over
// any shorter label another entry might share a substring with, and each
// matched span is consumed so the same text isn't double-counted.
function findObjectMentions(text: string, candidates: { typeId: string; label: string }[]): { typeId: string; index: number }[] {
  const sorted = [...candidates].sort((a, b) => b.label.length - a.label.length);
  const found: { typeId: string; index: number }[] = [];
  const consumed: [number, number][] = [];
  for (const c of sorted) {
    const idx = text.indexOf(c.label);
    if (idx === -1) continue;
    const overlaps = consumed.some(([s, e]) => idx < e && idx + c.label.length > s);
    if (overlaps) continue;
    found.push({ typeId: c.typeId, index: idx });
    consumed.push([idx, idx + c.label.length]);
  }
  return found.sort((a, b) => a.index - b.index);
}

export function parseEditCommand(text: string, variant: LayoutVariant): ParsedEditCommand | null {
  const lower = text.toLowerCase().trim();
  if (!lower) return null;
  const verb = findVerb(lower);
  if (!verb) return null;

  const presentTypeIds = [...new Set(variant.objects.map((o) => o.typeId))];
  const candidates = presentTypeIds
    .map((typeId) => ({ typeId, label: (OBJECT_LIBRARY[typeId]?.label ?? typeId).toLowerCase() }))
    .filter((c) => c.label.length > 0);

  const mentions = findObjectMentions(lower, candidates);
  if (mentions.length === 0) return null;

  const subjectTypeId = mentions[0].typeId;
  const referenceTypeId = mentions.length > 1 ? mentions[1].typeId : undefined;
  if ((verb === 'move-near' || verb === 'move-away') && !referenceTypeId) return null;

  return { verb, subjectTypeId, referenceTypeId };
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}

export function resizeTransform(entry: { minWidth: number; minHeight: number; defaultWidth: number; defaultHeight: number }, current: Transform, grow: boolean): { width: number; height: number } {
  const factor = grow ? 1.2 : 0.8;
  return {
    width: clamp(current.width * factor, entry.minWidth, entry.defaultWidth * 2.5),
    height: clamp(current.height * factor, entry.minHeight, entry.defaultHeight * 2.5),
  };
}

// Scans the plot for the best non-overlapping spot for `subject` (keeping
// its own size/rotation) that's as close to (mode 'near') or as far from
// (mode 'away') `reference' as the plot allows — the same coarse grid-scan
// approach placement.ts uses, but single-purpose: no sun/adjacency/setback
// scoring, since a manual quick-edit move is meant to behave like dragging
// the object by hand, not like re-running full generation.
export function findRepositionTarget(variant: LayoutVariant, plot: Plot, subject: PlanObject, reference: PlanObject, mode: 'near' | 'away'): Transform | null {
  const bounds = polygonBounds(plot.boundary);
  const w = subject.transform.width;
  const h = subject.transform.height;
  const step = Math.max(0.5, Math.min(bounds.maxX - bounds.minX, bounds.maxY - bounds.minY) / 40);
  const others = variant.objects.filter((o) => o.id !== subject.id);

  let best: { transform: Transform; score: number } | null = null;
  for (let x = bounds.minX + w / 2; x <= bounds.maxX - w / 2; x += step) {
    for (let y = bounds.minY + h / 2; y <= bounds.maxY - h / 2; y += step) {
      const transform: Transform = { x, y, width: w, height: h, rotationDeg: subject.transform.rotationDeg };
      if (!rectFullyInsidePolygon(transform, plot.boundary)) continue;
      const aabb = transformAabb(transform);
      if (others.some((o) => aabbOverlap(aabb, transformAabb(o.transform), 0.3))) continue;
      const d = distance(transform, reference.transform);
      const score = mode === 'near' ? -d : d;
      if (!best || score > best.score) best = { transform, score };
    }
  }
  return best?.transform ?? null;
}

File: swift/Sources/HomesteadEngine/EditCommands.swift
import Foundation

/// Bounded-vocabulary quick-edit commands, ported from
/// `src/engine/editCommands.ts` — the same spirit as `TextParser`'s brief
/// extraction: free text is matched against a fixed set of verbs and the
/// object labels actually present in the current plan, rather than sent to
/// an open-ended model, so results stay deterministic and explainable.
/// "Present in the plan" is correct here, not a bug to fix: every verb below
/// edits an object that has to already exist to be moved, resized, deleted,
/// or locked. (Adding a new object is a different action entirely —
/// `ObjectPalette`'s job, not this file's.)
public enum EditVerb: String, Codable, Sendable {
    case moveNear = "move-near"
    case moveAway = "move-away"
    case enlarge
    case shrink
    case delete
    case duplicate
    case rotate
    case lock
    case unlock
}

public struct ParsedEditCommand: Equatable, Sendable {
    public let verb: EditVerb
    public let subjectTypeId: String
    public let referenceTypeId: String?
}

public enum EditCommands {
    /// Order matters: more specific phrases ("away from") are tested before
    /// broader ones that would otherwise win first, and delete/duplicate/
    /// rotate/lock take priority over resize or move phrasing that might
    /// incidentally share a word.
    private static let verbPatterns: [(verb: EditVerb, pattern: String)] = [
        (.delete, #"\b(delete|remove)\b"#),
        (.duplicate, #"\b(duplicate|copy|clone)\b"#),
        (.rotate, #"\b(rotate|turn)\b"#),
        (.unlock, #"\bunlock\b"#),
        (.lock, #"\block\b"#),
        (.enlarge, #"\b(bigger|larger|enlarge|increase|grow|expand)\b"#),
        (.shrink, #"\b(smaller|shrink|decrease|reduce|shrink)\b"#),
        (.moveAway, #"\b(away from|farther from|further from|far from)\b"#),
        (.moveNear, #"\b(near|closer to|next to|beside|by the)\b"#),
    ]

    private static func findVerb(_ text: String) -> EditVerb? {
        for (verb, pattern) in verbPatterns {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { continue }
            if regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)) != nil { return verb }
        }
        return nil
    }

    /// Longest-label-first so a multi-word label ("Solar Panels") wins over
    /// any shorter label another entry might share a substring with; each
    /// matched span is consumed so the same text isn't double-counted.
    private static func findObjectMentions(_ text: String, candidates: [(typeId: String, label: String)]) -> [(typeId: String, index: Int)] {
        let sorted = candidates.sorted { $0.label.count > $1.label.count }
        var found: [(typeId: String, index: Int)] = []
        var consumed: [(Int, Int)] = []
        for candidate in sorted {
            guard let range = text.range(of: candidate.label) else { continue }
            let start = text.distance(from: text.startIndex, to: range.lowerBound)
            let end = start + candidate.label.count
            let overlaps = consumed.contains { start < $0.1 && end > $0.0 }
            if overlaps { continue }
            found.append((candidate.typeId, start))
            consumed.append((start, end))
        }
        return found.sorted { $0.index < $1.index }
    }

    public static func parse(_ text: String, objectsPresent: [PlanObject]) -> ParsedEditCommand? {
        let lower = text.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        guard !lower.isEmpty, let verb = findVerb(lower) else { return nil }

        var seenTypeIds = Set<String>()
        var presentTypeIds: [String] = []
        for object in objectsPresent where seenTypeIds.insert(object.typeId).inserted {
            presentTypeIds.append(object.typeId)
        }
        let candidates = presentTypeIds
            .map { (typeId: $0, label: (ObjectLibrary[$0]?.label ?? $0).lowercased()) }
            .filter { !$0.label.isEmpty }

        let mentions = findObjectMentions(lower, candidates: candidates)
        guard let first = mentions.first else { return nil }

        let referenceTypeId = mentions.count > 1 ? mentions[1].typeId : nil
        if (verb == .moveNear || verb == .moveAway), referenceTypeId == nil { return nil }

        return ParsedEditCommand(verb: verb, subjectTypeId: first.typeId, referenceTypeId: referenceTypeId)
    }

    public static func resizeTransform(entry: ObjectLibrary.Entry, current: Transform, grow: Bool) -> Size {
        let factor = grow ? 1.2 : 0.8
        return Size(
            width: clamp(current.width * factor, entry.minimumSize.width, entry.defaultSize.width * 2.5),
            height: clamp(current.height * factor, entry.minimumSize.height, entry.defaultSize.height * 2.5)
        )
    }

    public enum RepositionMode {
        case near, away
    }

    /// Scans the plot for the best non-overlapping spot for `subject`
    /// (keeping its own size/rotation) that is as close to (`.near`) or as
    /// far from (`.away`) `reference` as the plot allows — the same coarse
    /// grid-scan `Placement` uses, but single-purpose: no sun/adjacency/
    /// setback scoring, since a quick-edit move should behave like dragging
    /// the object by hand, not like re-running generation.
    public static func findRepositionTarget(objects: [PlanObject], plot: Plot, subject: PlanObject, reference: PlanObject, mode: RepositionMode) -> Transform? {
        guard let bounds = plot.bounds else { return nil }
        let w = subject.transform.width
        let h = subject.transform.height
        let step = max(0.5, min(bounds.width, bounds.height) / 40)
        let others = objects.filter { $0.id != subject.id }

        var best: (transform: Transform, score: Double)?
        var x = bounds.minX + w / 2
        while x <= bounds.maxX - w / 2 {
            var y = bounds.minY + h / 2
            while y <= bounds.maxY - h / 2 {
                let transform = Transform(x: x, y: y, width: w, height: h, rotationDeg: subject.transform.rotationDeg)
                if Polygon.contains(transform, polygon: plot.boundary) {
                    let aabb = transform.aabb
                    let overlapsAny = others.contains { aabb.overlaps($0.transform.aabb, margin: 0.3) }
                    if !overlapsAny {
                        let d = distance(transform.center, reference.transform.center)
                        let score = mode == .near ? -d : d
                        if best == nil || score > best!.score { best = (transform, score) }
                    }
                }
                y += step
            }
            x += step
        }
        return best?.transform
    }
}
